modules/rest/src/main/java/org/apache/ignite/rest/presentation/json/JsonConverter.java

```diff
@@ -35,6 +52,213 @@
         return gson.toJson(obj);
     }
 
+    /** */
+    public static ConfigurationVisitor<JsonElement> jsonVisitor() {
+        return new ConfigurationVisitor<JsonElement>() {
+            /** */
+            private final Deque<JsonObject> jsonObjectsStack = new ArrayDeque<>();
+
+            /** {@inheritDoc} */
+            @Override public JsonElement visitLeafNode(String key, Serializable val) {
+                JsonElement jsonLeaf = toJsonLeaf(val);
+
+                if (!jsonObjectsStack.isEmpty())
+                    jsonObjectsStack.peek().add(key, jsonLeaf);
+
+                return jsonLeaf;
+            }
+
+            /** {@inheritDoc} */
+            @Override public JsonElement visitInnerNode(String key, InnerNode node) {
+                JsonObject innerJsonNode = new JsonObject();
+
+                jsonObjectsStack.push(innerJsonNode);
+
+                node.traverseChildren(this);
+
+                jsonObjectsStack.pop();
+
+                if (!jsonObjectsStack.isEmpty())
+                    jsonObjectsStack.peek().add(key, innerJsonNode);
+
+                return innerJsonNode;
+            }
+
+            /** {@inheritDoc} */
+            @Override public <N extends InnerNode> JsonElement visitNamedListNode(String key, NamedListNode<N> node) {
+                JsonObject namedListJsonNode = new JsonObject();
+
+                jsonObjectsStack.push(namedListJsonNode);
+
+                for (String subkey : node.namedListKeys())
+                    node.get(subkey).accept(subkey, this);
+
+                jsonObjectsStack.pop();
+
+                if (!jsonObjectsStack.isEmpty())
+                    jsonObjectsStack.peek().add(key, namedListJsonNode);
+
+                return namedListJsonNode;
+            }
+
+            /** */
+            @NotNull private JsonElement toJsonLeaf(Serializable val) {
+                if (val == null)
+                    return JsonNull.INSTANCE;
+
+                Class<? extends Serializable> valClass = val.getClass();
+
+                if (!valClass.isArray())
+                    return toJsonPrimitive(val);
+
+                JsonArray jsonArray = new JsonArray();
+
+                for (int i = 0; i < Array.getLength(val); i++)
+                    jsonArray.add(toJsonPrimitive(Array.get(val, i)));
+
+                return jsonArray;
+            }
+
+            /** */
+            @NotNull private JsonElement toJsonPrimitive(Object val) {
+                if (val == null)
+                    return JsonNull.INSTANCE;
+
+                if (val instanceof Boolean)
+                    return new JsonPrimitive((Boolean)val);
+
+                if (val instanceof String)
+                    return new JsonPrimitive((String)val);
+
+                if (val instanceof Number)
+                    return new JsonPrimitive((Number)val);
+
+                assert false : val;
```

Comment:
what's the point of this assertion?

Reply:
This branch must be impossible, all types are already listed in corresponding conditional expressions.

Reply:
But you throw an exception below anyway, so you would get the same result. Or am I missing something?

Reply:
That's because assertions can be disabled, I have to put something in this default branch so compiler would compile it.

Reply:
I understand that, but why do you need the assertion if you throw an exception in this case anyway? The only difference is that you would get a different exception when assertions are enabled...

Reply:
The difference is that "assert false" explicitly says that this code branch is not valid